lib/rest-adapter.js

```diff
@@ -403,6 +411,14 @@ RestAdapter.prototype._registerMethodRouteHandlers = function(router,
     // Express 4.x only supports delete
     verb = 'delete';
   }
+
+  //For now: exclude paths with parameters like :id
+  //and handler functions with err obj i.e. (err, req,res,next)
+  if (!route.path.includes(':') && handler.length < 4) {
+    handler.methodName = sharedMethod.name;
+    handler.path = route.fullPath;
+    customRouter.registerPathAndHandlers({ verb: verb, fullPath: route.fullPath }, handler);
+  }
   router[verb](route.path, handler);
```

Comment:
I could not find this comment on GH, so I am responding here.
> the built-in router places all verb handler on the top of the handler stack, and keeps a flag to check if the path has an all handler, if yes, it > invokes the all handler first and runs other registered handlers> In our case, trie has a placeholder for all handler, and we can check, but we are only inovking single handler.
That makes perfect sense to me, now we just need to improve error logging so that the user understands what's happening under the hood.
 
1) If there is an "all" handler already registered for a path, and users is attempting to add another handler, then we should do the same thing we do now when the second handler is registered for the same "path+verb"
 
2) When registering an "all" handler, check if there are any existing verb-specific handlers for the same path. If there are, then report a warning for each of them, since they won't be invoked.
 
BTW are you sure the "all" verb handler is on the top? I would expect it's at the bottom, i.e. more specific handlers (GET, POST) are run before generic ones (ALL).
> I'm thinking of changing it in a way that:> in all: handler(req,res,next) , next points to the second handler in the stack for the path and so on.. I will commit this approach in the next commit.> Or, should you have any better idea for this scenario -- after looking at recent commit -- please advise. Thank you :)
AFAIK, all strong-remoting handler functions either send back the response (and never call next), or fail with an error and call next(err). In both cases the flow control skips all subsequent handler functions.
 
This is why we keep only the first handler for each path+verb (instead of an array), and that's why I believe we don't need to chain handlers for "all" verb either. Am I missing something?
 
Miroslav
 
 
----- Original message -----From: Gunjan Pandya <notifications@github.com>To: strongloop/strong-remoting <strong-remoting@noreply.github.com>Cc: Miroslav Bajtos/Czech Republic/IBM@IBMCZSubject: Re: [strong-remoting] Implement a faster REST router (#282)Date: Mon, Feb 8, 2016 9:48 AM 
In lib/rest-adapter.js:
> @@ -403,6 +411,14 @@ RestAdapter.prototype._registerMethodRouteHandlers = function(router,>      // Express 4.x only supports delete>      verb = 'delete';>    }> +> +  //For now: exclude paths with parameters like :id> +  //and handler functions with err obj i.e. (err, req,res,next)> +  if (!route.path.includes(':') && handler.length < 4) {> +    handler.methodName = sharedMethod.name;> +    handler.path = route.fullPath;> +    customRouter.registerPathAndHandlers({ verb: verb, fullPath: route.fullPath }, handler);> +  }>    router[verb](route.path, handler);
@bajtos : the built-in router places all verb handler on the top of the handler stack, and keeps a flag to check if the path has an all handler, if yes, it invokes the all handler first and runs other registered handlersIn our case, trie has a placeholder for all handler, and we can check, but we are only inovking single handler.I'm thinking of changing it in a way that:in all: handler(req,res,next) , next points to the second handler in the stack for the path and so on.. I will commit this approach in the next commit.
Or, should you have any better idea for this scenario -- after looking at recent commit -- please advise. Thank you :)
—Reply to this email directly or view it on GitHub.

 